cookbook/playground/gemini_agents.py

```diff
@@ -0,0 +1,16 @@
+from phi.agent import Agent
+from phi.tools.yfinance import YFinanceTools
+from phi.playground import Playground, serve_playground_app
+from phi.model.google import Gemini
+
+finance_agent = Agent(
+    name="Finance Agent",
+    model=Gemini(id="gemini-2.0-flash-exp"),
+    tools=[YFinanceTools(stock_price=True)],
+    debug_mode=True,
+)
```

Comment:
Add these two flags
- `add_chat_history_to_messages=True,`
- `num_history_responses=3`

And see if you get error.

Persistent memory is not working.

Reply:
please test with this sample config

  ``` 
master = Agent(
      name="Master",
      model=model,
      agent_id="master",
      team=[web_agent],
      storage=SqlAgentStorage(table_name="master_sessions", db_file="tmp/agents.db"),
      markdown=True,
      add_chat_history_to_messages=True,
      num_history_responses=3
  )
```

Reply:
@samyogdhital @manthanguptaa It seems that persistence still cannot save messages related to tool use in Gemini
Is there a fix on the way?
If not, I can try to fix it after the New Year holidays.


```
DEBUG    ============== user ==============                                                                                                                                
DEBUG    how many messages did i send?                                                                                                                                     
DEBUG    ============== model ==============                                                                                                                               
DEBUG    Tool Calls: [                                                                                                                                                     
           {                                                                                                                                                               
             "type": "function",                                                                                                                                           
             "function": {                                                                                                                                                 
               "name": "get_chat_history",                                                                                                                                 
               "arguments": "{}"                                                                                                                                           
             }                                                                                                                                                             
           }                                                                                                                                                               
         ]                                                                                                                                                                 
DEBUG    ============== tool ==============                                                                                                                                
DEBUG    ['']                                                                                                                                                              
DEBUG    ============== model ==============                                                                                                                               
DEBUG    I'm sorry, I cannot fulfill this request. There is no information about how many messages you sent.                                                               
                                                                                                                                                                           
DEBUG    **************** METRICS START ****************                                                                                                                   
DEBUG    * Time to first token:         1.4130s                                                                                                                            
DEBUG    * Time to generate response:   1.5110s                                                                                                                            
DEBUG    * Tokens per second:           15.2219 tokens/s                                                                                                                   
DEBUG    * Input tokens:                1065                                                                                                                               
DEBUG    * Output tokens:               23                                                                                                                                 
DEBUG    * Total tokens:                1088                                                                                                                               
DEBUG    **************** METRICS END ******************                                                                                                                   
DEBUG    ---------- Gemini Response End ----------                                                                                                                         
DEBUG    ---------- Gemini Response End ----------                                                                                                                         
DEBUG    Added 4 Messages to AgentMemory                                                                                                                                   
DEBUG    Added AgentRun to AgentMemory                                                                                                                                     
DEBUG    Exception upserting into table: (builtins.TypeError) Object of type RepeatedComposite is not JSON serializable                                                    
         [SQL: INSERT INTO agent_sessions (session_id, agent_id, user_id, memory, agent_data, user_data, session_data, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?) ON      
         CONFLICT (session_id) DO UPDATE SET agent_id = ?, user_id = ?, memory = ?, agent_data = ?, user_data = ?, session_data = ?, updated_at = ?]                       
         [parameters: [{}]]                                                                                                                                                
DEBUG    Checking if table exists: agent_sessions                                                                                                                          
DEBUG    *********** Agent Run End: df9761df-1868-4b4d-afc4-b7305132d7b8 ***********                                                                                       

```


 **Exception upserting into table: (builtins.TypeError) Object of type RepeatedComposite is not JSON serializable**